[Clang][HLSL] Add environment parameter to availability attribute

hekota opened the pull request:
Add `environment` parameter to Clang availability attribute. The allowed values for this parameter are a subset of values allowed in the `llvm::Triple` environment component. If the `environment` parameters is present, the declared availability attribute applies only to targets with the same platform and environment.

This new parameter will be initially used for annotating HLSL functions for the `shadermodel` platform because in HLSL built-in function availability can depend not just on the shader model version (mapped to `llvm::Triple::OSType`) but also on the target shader stage (mapped to `llvm::Triple::EnvironmentType`). See example in #89802 and microsoft/hlsl-specs#204 for more details.

The environment parameter is currently supported only for HLSL.

Fixes #89802

hekota committed:
Add environment parameter to clang availability attribute

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into availability-environment-param

hekota committed:
apply clang-format

hekota committed:
Add tests and documentation, adjust diag messages

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into availability-environment-param

llvm-beanz reviewed (commented):
Overall I really like the direction of this. I'm curious if any of the maintainers from Apple have thoughts since they're the primary users of availability annotations.

I think this dramatically simplifies things for what we need, but it might also have some benefits to be able to simplify how Apple expresses some of its existing annotations.

(cc @cyndyishida 👋 )

hekota commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3859,7 +3862,7 @@ def warn_availability_fuchsia_unavailable_minor : Warning<
 
 def warn_unguarded_availability :
-  Warning<"%0 is only available on %1 %2 or newer">,
+  Warning<"%0 %select{is only|is not}5 available %select{|in %4 environment }3on %1 %2 %select{or newer|}5">,
```
If the API is available in higher version the warning will always be:

"%0 is only available on %1 %2 or newer"

The other formulation is only used when the API is not available in any future version for the target environment. This happens when there is an attribute narrowing down the availability to specific environment that is different from the current one. I kept the message under one ID so it can be turned on/off with `-Wunguarded-availability-new`.

hekota replied:
I have modified the diagnostic message when the API is not available in the target environment for any version to say that it is unavailable and added a note with additional information like this:

```
warning: 'A' is unavailable
note: 'A' has been marked as being introduced in %attr-platform %attr-version %attr-env environment here, but the deployment target is %target-platform %target -version %target-env environment.
```

Endilll reviewed (commented):
`Sema.h` changes look good.

python3kgae commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -1016,7 +1016,34 @@ static llvm::StringRef canonicalizePlatformName(llvm::StringRef Platform) {
+             .Default(Environment);
+}
+static llvm::Triple::EnvironmentType getEnvironmentType(llvm::StringRef Environment) {
+    return llvm::StringSwitch<llvm::Triple::EnvironmentType>(Environment)
```
Will this support environment other than HLSL profiles?

hekota replied:
Other environments can be easily added as needed.

cyndyishida commented:
> Overall I really like the direction of this. I'm curious if any of the maintainers from Apple have thoughts since they're the primary users of availability annotations.

While I like the approach of aligning availability parameters closer to `llvm::Triple`, I am concerned about how this will interact with existing precedent. There is a lot of code that passes in _platform_ values that aren't actually `OSType`. For example `__attribute__((availability(macCatalyst, ...))` aligns to `EnviornmentType::MacABI` and 
`__attribute__((availability(macOSApplicationExtension, ...)))` aligns to  cli args `-mtargetos=macos -fapplication-extension`. It may become difficult to maintain the distinction between what is accepted as a `platform` vs a `environment`. And we can't realistically re-write client code relying on this. This attribute behavior also aligns very closely with Swift's `@available`
Would it be possible to similarly treat the target shader stages & combinations as platform values as well? Or define a new attribute that does not conflict.

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into availability-environment-param

hekota committed:
Remove trailing whitespace

hekota committed:
Modify diagnostic message when function is not available for target environment for any version

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into availability-environment-param

hekota committed:
Move tests into Sema, add coverage for strict flag

- merge AvailabilityMarkup.hlsl and AvailabilityMarkupStages.hlsl into attr-availability-shadermodel.hlsl and move to Sema test directory
- add coverage for strict mode which uses different code path to generate the warnings
- add missing space in diag message

hekota commented:
Ping

hekota committed:
Fix test - add numthreads attribute

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into availability-environment-param

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into availability-environment-param

python3kgae commented on clang/lib/Headers/hlsl/hlsl_intrinsics.h:
```diff
@@ -18,14 +18,21 @@ namespace hlsl {
-  __attribute__((availability(environment, introduced = version)))
+#define _HLSL_16BIT_AVAILABILITY(platform, version)                            \
+  __attribute__((availability(platform, introduced = version)))
+#define _HLSL_16BIT_AVAILABILITY_STAGE(platform, version, stage)               \
```
Is there anything that only support 16bit on special stages ?

hekota replied:
I don't think so. The change in this file is just about renaming the macro argument from `environment` to `platform` so it does not get confused with the new `environment` parameter on the attribute.

llvm-beanz commented on clang/lib/Sema/SemaAvailability.cpp:
```diff
@@ -376,26 +400,46 @@ static void DoEmitAvailabilityWarning(Sema &S, AvailabilityResult K,
 
-    S.Diag(Loc, Warning) << OffendingDecl << PlatformName
-                         << Introduced.getAsString();
+    if (EnvironmentMatchesOrNone) {
+      unsigned DiagKind = UseNewWarning ? diag::warn_unguarded_availability_new
+                                        : diag::warn_unguarded_availability;
+
+      S.Diag(Loc, DiagKind)
+          << OffendingDecl << PlatformName << Introduced.getAsString()
+          << UseEnvironment << TargetEnvironment;
+    } else {
+      unsigned DiagKind =
+          UseNewWarning ? diag::warn_unguarded_availability_unavailable_new
+                        : diag::warn_unguarded_availability_unavailable;
+      S.Diag(Loc, DiagKind) << Loc << OffendingDecl;
+    }
```
What about something like this?
```suggestion
    unsigned Warning =
        EnvironmentMatchesOrNone
            ? (UseNewWarning ? diag::warn_unguarded_availability_new
                             : diag::warn_unguarded_availability)
            : (UseNewWarning ? diag::warn_unguarded_availability_unavailable_new
                             : diag::warn_unguarded_availability_unavailable);
    S.Diag(Loc, DiagKind) << OffendingDecl << PlatformName
                          << Introduced.getAsString() << UseEnvironment
                          << TargetEnvironment;
```

The the new `warn_unguarded_availability_unavailable` has strictly a subset of the arguments for the other diagnostic so we should be able to just substitute them like this.

llvm-beanz reviewed (commented):
Overall this looks good to me. One set of suggestions inline.

farzonl commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3861,11 +3864,17 @@ def warn_availability_fuchsia_unavailable_minor : Warning<
 
 def warn_unguarded_availability :
-  Warning<"%0 is only available on %1 %2 or newer">,
+  Warning<"%0 is only available %select{|in %4 environment }3on %1 %2 or newer">,
```
is "3on" a typo? seems like it should be " on".

farzonl commented on clang/lib/Headers/hlsl/hlsl_intrinsics.h:
```diff
@@ -18,14 +18,21 @@ namespace hlsl {
-  __attribute__((availability(environment, introduced = version)))
+#define _HLSL_AVAILABILITY(platform, version)                                  \
+  __attribute__((availability(platform, introduced = version)))
+#define _HLSL_AVAILABILITY_STAGE(platform, version, stage)                     \
```
should we be updating the below intrinsics to include shader stages?

hekota replied:
I believe all of the intrinsics implemented so far are available for all shader stages, so no update is needed.

hekota committed:
Code review feedback

farzonl commented on clang/test/Sema/attr-availability-shadermodel.hlsl:
```diff
@@ -0,0 +1,140 @@
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel5.0-compute -fsyntax-only -verify %s
```
could we do different runs for pixel and mesh to cover their different version availabilities from compute?

hekota replied:
Done.

hekota committed:
Add more tests, disable availability strict parameter use with environment

hekota committed:
clang-format

hekota committed:
Address feedback - restricting environment parameter to HLSL only and prohibiting strict flag in HLSL

llvm-beanz commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3859,13 +3862,25 @@ def note_protocol_method : Note<
 def warn_availability_fuchsia_unavailable_minor : Warning<
   "Fuchsia API Level prohibits specifying a minor or sub-minor version">,
   InGroup<Availability>;
+def warn_availability_hlsl_unavailable_strict: Warning<
+  "strict parameter is not supported in HLSL">,
+  InGroup<Availability>;
+def warn_availability_environment_only_in_hlsl: Warning<
+  "environment parameter is supported only in HLSL">,
+  InGroup<Availability>;
```
It seems to me like these two should be errors not warnings, and maybe we can collapse them down. Something like:
```suggestion
def err_availability_unexpected_parameter: Error<
  "unexpected parameter %0 in availability attribute, not permitted in %select{HLSL|C/C++}1">;
```

hekota committed:
Code review feedback - unify diag message and make it an error

hekota committed:
One more test update

llvm-beanz commented on clang/test/Sema/attr-availability-shadermodel-compute.hlsl:
nit: we've generally tried to keep the HLSL-specific tests separated from the other language tests (i.e. SemaHLSL).

hekota replied:
Right, and I am adding a lot more HLSL-specific tests in my upcoming PRs that implement the diagnostic modes under the SemaHLSL directory. There is a number of platform-specific tests for the availability attribute already here under Sema though so I figured we should have some basic coverage here as well.

llvm-beanz replied:
I think the different slice there is platform vs language. SemaHLSL should be where HLSL tests go regardless of platform (DirectX or Vulkan), but we should try to keep HLSL separate from C/C++ tests.

llvm-beanz reviewed (commented):
A couple comments, but mostly this looks good to me.

hekota committed:
Move tests under SemaHLSL
